### PR TITLE
[Feature] Incorporate organization in APIM workflows

### PR DESCRIPTION
**Pull request by Sarangan0219**
## Purpose

This is regarding the effort to support the organization concept in the non-registry side for APIM.

APIM will receive the organization parameter (organization UUID) as a query parameter in the REST API layer. Necessary validations related to that organization UUID will be done in the REST API layer and it will be passed to the workflow if the validation succeeds.

We will be using the tenantDomain or OrganizationUUID in the registry layer to create and maintain respective collections. Thus We need to propagate these parameters from the REST API layer to Registry Layer.

To cater to all deployments, We will be passing an **organizationID** parameter from the REST API layer to the Registry layer. Most of the methods are having tenantDomain as a method parameter in them. Tenant Domain and organaizationUUID will be used as the organizationID. By this approach, We don’t need to change the method signature of existing methods, which has tenantDomain as a parameter. We just need to rename the parameter. For other methods where the tenant domain is not a parameter, We will be adding an extra parameter of organizationID in the base APIManager, APIProvider, and API COnsumer  Interfaces and propagate that to all affected methods. Through this effort, we will be covering most of the Publisher and Store functionalities.

## Implementation

We will be following the below-mentioned approaches in this effort:

- Replace **orgId** object instead of **tenantDomain** parameter.

- Replace **Organization** object instead of **requestedTenantDomain** parameter. In an operation flow where the tenantDomain is requested by the user then we need to populate that to the workflow together with the organizationId. Thus an Organization DTO is being populated and retrieved at required Places.

- Create New methods with ** organizationID** as a parameter to perform DB operations. Implemented separate DB operations in DAO layer.

- Add **orgId** as a new parameter.

- For the methods which have an API object as a parameter, set the **organizationId** as an attribute to the API object.

### Methods where  organizationID is added as a parameter

**APIManager.java**

- List<String>getApiVersionsMatchingApiNameAndOrganization(String apiName,String username, String organizationId)

### Methods where orgId is added as a parameter

**APIManager.java**

- boolean isApiNameExist(String apiName, String orgId) 
- boolean isApiNameWithDifferentCaseExist(String apiName, String orgId)
- String getOpenAPIDefinition(Identifier apiId, String orgId)
- String getOpenAPIDefinition(String apiId, String orgId)
- boolean isDocumentationExist(String uuid, String docName, String orgId)

**APIProvider.java**

- void removeDocumentation(String apiId, String documentId, String orgId)
- void removeDocumentation(Identifier id, String docId, String orgId)
- Documentation addDocumentation(String uuid, Documentation documentation, String orgId)
- void addDocumentationContent(String uuid, String docId, String orgId, DocumentationContent content)
- Documentation updateDocumentation(String apiId, Documentation documentation, String orgId)
- void saveSwagger20Definition(APIIdentifier apiId, String jsonText, String orgId)
- void saveSwaggerDefinition(API api, String jsonText, String orgId
- void saveSwaggerDefinition(String apiId, String jsonText, String orgId)
- APIStateChangeResponse changeLifeCycleStatus(String orgId, String uuid, String action, Map<String, Boolean> checklist)
- Map<String, Object> getAPILifeCycleData(String apiId, String orgId)
- void addWSDLResource(String apiId, ResourceFile resource, String url, String OrgId)
- void setThumbnailToAPI(String apiId, ResourceFile resource, String orgId)
- List<Mediation> getAllApiSpecificMediationPolicies(String apiId, String orgId)
- Mediation getApiSpecificMediationPolicyByPolicyId(String apiId, String policyId, String orgId)
- Mediation addApiSpecificMediationPolicy(String apiId, Mediation mediationPolicy, String orgId)
- Mediation updateApiSpecificMediationPolicyContent(String apiId, Mediation mediationPolicy, String orgId)
- void deleteApiSpecificMediationPolicy(String apiId, String mediationPolicyId, String orgId)
- void saveGraphqlSchemaDefinition(String apiId, String definition, String orgId)

**APIConsumer.java**

- getLightweightAPI(APIIdentifier identifier, String orgId)
- ResourceFile getWSDL(API api, String environmentName, String environmentType, String orgId)

### Methods where orgId is replaced for tenantDomain

**APIManager.java**

- String getGraphqlSchemaDefinition(String apiId, String orgId)
- List<Documentation> getAllDocumentation(String uuid, String orgId)
- Documentation getDocumentation(String apiId, String docId, String orgId)
- DocumentationContent getDocumentationContent(String apiId, String docId, String orgId)
- ResourceFile getIcon(String apiId, String orgId)
- Map<String, Object> searchPaginatedAPIs(String searchQuery, String orgId, int start, int end, boolean limitAttributes, boolean isPublisherListing)
- Map<String, Object> searchPaginatedContent(String searchQuery, String orgId, int start, int end)
- Map<String, Object> searchPaginatedContent(String searchQuery, String orgId, int start, int end)

### Methods where Organization Object is replaced for tenantDomain

**APIManager.java**

- API getAPIbyUUID(String uuid, String orgId)
- ApiTypeWrapper getAPIorAPIProductByUUID(String uuid, String orgId)
- API getLightweightAPIByUUID(String uuid, String orgId)
- API getLightweightAPI(APIIdentifier identifier, String orgId)

### Methods where API object is passed with **organizationId**

- API updateAPI(API api, API existingAPI)
- API addAPI(API api)

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 1]
```diff
@@ -797,11 +804,18 @@ public static API prepareToCreateAPIByDTO(APIDTO body, APIProvider apiProvider,
...
-        if (apiProvider.isApiNameWithDifferentCaseExist(body.getName())) {
-            throw new APIManagementException(
-                    "Error occurred while adding API. API with name " + body.getName() + " already exists.",
-                    ExceptionCodes.API_ALREADY_EXISTS);
+
+
+        if (organizationId != null) {
+
+        } else {
+            if (apiProvider.isApiNameWithDifferentCaseExist(body.getName())) {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:812]: can we combine these into one If?

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/dto/OrganizationDTO.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package org.wso2.carbon.apimgt.api.dto;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/dto/OrganizationDTO.java:1]: Missing License header

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/dto/OrganizationDTO.java:20]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/adb1f3dbfaedb3a5a796cfdaab55385b2e4dfb1b

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java (hunk 1 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/adb1f3dbfaedb3a5a796cfdaab55385b2e4dfb1b

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java [hunk 3]
```diff
@@ -18,12 +18,12 @@
 
 package org.wso2.carbon.apimgt.api;
 
+import org.wso2.carbon.apimgt.api.dto.OrganizationDTO;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java:21]: shall we move this to model package?

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java:21]: I think right place for Organization (in persistence package) is api package. we can move it later

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java [hunk 4]
```diff
@@ -877,24 +899,25 @@ String addProductResourceFile(APIProductIdentifier identifier, String resourcePa
...
-     * @param tenantDomain tenant domain
+     * @param orgId Identifier of an organization
      * @param start starting number
      * @param end ending number
      * @return
      * @throws APIManagementException
      */
-    Map<String, Object> searchPaginatedAPIs(String searchQuery, String tenantDomain, int start, int end)
+    Map<String, Object> searchPaginatedAPIs(String searchQuery, String orgId, int start,
+                                            int end)
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java:909]: not required to change this line?

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java [hunk 5]
```diff
@@ -357,6 +354,14 @@
      */
     void manageAPI(API api) throws APIManagementException, FaultGatewaysException;
 
+    /**
+     * Returns the UUID of the organization which the API belongs to
+     * @param apiId UUID of the API
+     * @return Organization UUID of the given artifact id or null
+     * @throws APIManagementException if failed get API from APIIdentifier
+     */
+    String getOrganizationIDbyAPIUUID(String apiId) throws APIManagementException;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java:363]:
```suggestion
    String getOrganizationIdByAPIUUID(String apiId) throws APIManagementException;
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 6]
```diff
@@ -977,8 +992,13 @@ public API addAPI(API api) throws APIManagementException {
...
 
         try {
-            PublisherAPI addedAPI = apiPersistenceInstance.addAPI(new Organization(tenantDomain),
-                    APIMapper.INSTANCE.toPublisherApi(api));
+            Organization org;
+            if (api.getOrganizationId() != null) {
+                org = new Organization(api.getOrganizationId());
+            } else {
+                org = new Organization(tenantDomain);
+            }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:1000]: How about setting an OrganizationDTO itself into API object.
The behaviour would be:
Choreo Org case:

OrganizationDTO.getOrgID = <choreo_org>
OrganizationDTO.getTenantDomain = carbon.super

On-premise

OrganizationDTO.getOrgID = <tenantDomain of API>
OrganizationDTO.getTenantDomain = <tenantDomain of API>

Then we can just get API.getOrganizationDTO().getOrgID() and assign to this.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 7]
```diff
@@ -1778,8 +1798,14 @@ public API updateAPI(API api, API existingAPI) throws APIManagementException, Fa
                 }
             }
             try {
+                Organization org;
+                if (api.getOrganizationId() != null) {
+                    org = new Organization(api.getOrganizationId());
+                } else {
+                    org = new Organization(tenantDomain);
+                }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:1806]: could avoid the If block in the same way and make it a single line?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 8]
```diff
@@ -202,6 +201,95 @@ private ApiMgtDAO() {
...
+            ps.setString(2, organizationId);
+            resultSet = ps.executeQuery();
+            while (resultSet.next()) {
+                versionList.add(resultSet.getString("API_VERSION"));
+            }
+        } catch (SQLException e) {
+            handleException("Failed to get API versions matches API name" + apiName, e);
+        } finally {
+            APIMgtDBUtil.closeAllConnections(ps, conn, resultSet);
+        }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:225]: Shall we use same pattern as [1] in all the DB methods
1. Dealing with AutoCommit
2. Use try with statements
3. Rolling back during an issue


[1] https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ServiceCatalogDAO.java#L126-L162

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 9]
```diff
@@ -8065,7 +8166,8 @@ public APIProductIdentifier getAPIProductIdentifierFromUUID(String uuid) throws
     public String getUUIDFromIdentifier(APIIdentifier identifier) throws APIManagementException {
 
         String uuid = null;
-        String sql = SQLConstants.GET_UUID_BY_IDENTIFIER_SQL;
+
+        String sql  = SQLConstants.GET_UUID_BY_IDENTIFIER_SQL;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8170]: extra space

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 10]
```diff
@@ -2807,6 +2808,22 @@ public static APIStore getExternalAPIStore(String apiStoreName, int tenantId) th
...
+
+    public static OrganizationDTO getOrganizationDTOFromOrgID(String orgId, String requestedTenantDomain) {
+        OrganizationDTO organizationDTO = new OrganizationDTO(orgId);
+        if (requestedTenantDomain != null) {
+            organizationDTO.setRequestedTenantDomain(requestedTenantDomain);
+        } else {
+            organizationDTO.setRequestedTenantDomain(CarbonContext.getThreadLocalCarbonContext().getTenantDomain());
+        }
+        return  organizationDTO;
+    }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java:2825]:
```suggestion
    public static OrganizationDTO getOrganizationDTOFromOrgID(String orgId) {
        OrganizationDTO organizationDTO = new OrganizationDTO(orgId);
        organizationDTO.setRequestedTenantDomain(CARBON.SUPER);
        return  organizationDTO;
    }
```

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/dto/Organization.java [hunk 11]
```diff
@@ -23,19 +23,19 @@ public void setName(String name) {
...
-}
+
+    public String getOrgId() {
+        return orgId;
+    }
+
+    public void setOrgId(String orgId) {
+        this.orgId = orgId;
+    }
+}
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/dto/Organization.java:41]: Lets add a new line

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java [hunk 12]
```diff
@@ -5,17 +5,14 @@
 import org.joda.time.DateTime;
 import org.joda.time.format.DateTimeFormatter;
 import org.joda.time.format.ISODateTimeFormat;
-import org.wso2.carbon.apimgt.api.APIConsumer;
-import org.wso2.carbon.apimgt.api.APIDefinition;
-import org.wso2.carbon.apimgt.api.APIManagementException;
-import org.wso2.carbon.apimgt.api.APIMgtAuthorizationFailedException;
-import org.wso2.carbon.apimgt.api.APIProvider;
+import org.wso2.carbon.apimgt.api.*;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java:8]: Lets use single imports

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java [hunk 13]
```diff
@@ -397,6 +392,47 @@ public static void validateUserTenantWithAPIIdentifier(APIIdentifier apiIdentifi
         }
     }
 
+    /**
+     * Getting the Identifier of an Organization
+     * @param organizationId UUID of the Organization which the API belongs to
+     * @param tenantDomain   TenantDomain which the API belongs to
+     * @return tenantDomain or OrganizationUUID
+     */
+    public static String getOrgIdMatchesTenantDomain(String organizationId, String tenantDomain)
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java:401]: getMatchingOrgIdOrTenantDomain?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 14]
```diff
@@ -239,8 +231,8 @@ public Response getAllAPIs(Integer limit, Integer offset, String xWSO2Tenant, St
                 RestApiUtil.handleMigrationSpecificPermissionViolations(tenantDomain, username);
             }*/
             Map<String, Object> result;
-
-            result = apiProvider.searchPaginatedAPIs(query, tenantDomain, offset, limit);
+            String orgId = PublisherCommonUtils.getOrgId(organizationId);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:234]: are there multiple getOrgId methods?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 15]
```diff
@@ -2160,14 +2244,13 @@ public Response addAPIMediationPolicy(String apiId, String type, String ifMatch,
                 OMElement seqElement = APIUtil.buildOMElement(new ByteArrayInputStream(sequenceBytes));
                 String localName = seqElement.getLocalName();
                 fileName = seqElement.getAttributeValue(new QName("name"));
-                
                 if (APIConstants.MEDIATION_SEQUENCE_ELEM.equals(localName)) {
                     Mediation mediationPolicy = new Mediation();
                     mediationPolicy.setConfig(content);
-                    mediationPolicy.setName(fileName);
+                    mediationPolicy.setName(localName);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:2250]: This change done due to a different issue?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 16]
```diff
@@ -231,17 +251,23 @@ public Response apisApiIdGraphqlSchemaGet(String apiId, String ifNoneMatch, Stri
                 String errorMessage = "Error while retrieving API : " + apiId;
                 RestApiUtil.handleInternalServerError(errorMessage, e, log);
             }
+        } catch (UserStoreException e) {
+            String errorMessage = "Error while checking availability of tenant " + xWSO2Tenant;
+            RestApiUtil.handleInternalServerError(errorMessage, e, log);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java:256]: Looks like this is repeated in many places. Shall we use an exception code based fix. Can do this later if need.
https://github.com/wso2/carbon-apimgt/wiki/Exception-handling-in-REST-APIs

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 17]
```diff
@@ -710,24 +741,21 @@ public Response apisApiIdSwaggerGet(String apiId, String labelName, String envir
...
+            String orgId = RestApiCommonUtil.getOrgIdMatchesTenantDomain(organizationId, xWSO2Tenant);
             APIConsumer apiConsumer = RestApiCommonUtil.getLoggedInUserConsumer();
-            if (!APIUtil.isTenantAvailable(requestedTenantDomain)) {
-                RestApiUtil.handleBadRequest("Provided tenant domain '" + xWSO2Tenant + "' is invalid",
-                        ExceptionCodes.INVALID_TENANT.getErrorCode(), log);
-            }
             //this will fail if user does not have access to the API or the API does not exist
-            apiConsumer.getLightweightAPIByUUID(apiId, requestedTenantDomain);
-            ResourceFile thumbnailResource = apiConsumer.getIcon(apiId, requestedTenantDomain);
+            //APIIdentifier apiIdentifier = APIMappingUtil.getAPIIdentifierFromUUID(apiId, requestedTenantDomain);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java:757]: shall we remove the commented line if not needed?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 18]
```diff
@@ -908,83 +929,99 @@ public Response apisApiIdUserRatingDelete(String apiId, String xWSO2Tenant, Stri
...
+
+                return RestApiUtil.getResponseFromResourceFile(apiIdentifier.toString(), wsdl);
+            } else {
+                throw new APIManagementException(ExceptionCodes.from(ExceptionCodes.NO_GATEWAY_ENVIRONMENTS_ADDED,
+                        apiIdentifier.toString()));
+            }
+        } catch (UserStoreException e) {
+            String errorMessage = "Error while checking availability of tenant " +xWSO2Tenant;
+            RestApiUtil.handleInternalServerError(errorMessage, e, log);
         }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java:978]: I think its better not to throw a UserStoreException from RestApiCommonUtil.getOrgIdMatchesTenantDomain. You can catch UserStoreException inside that method and throw an APIManagementException with a proper [ExceptionCode](https://github.com/wso2/carbon-apimgt/wiki/Exception-handling-in-REST-APIs). Then just throw the APIManagementException from this method too. REST API can automatically handle it.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java (hunk 14 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:234]: Yes, there are 2 different **orgId** retrieval methods.

1. Publisher - In Publisher tenant domain is not coming from the user as a request parameter. We are getting it from the context and pass it to the workflow. So we only need to pass **organizationId** to the method logic.

**_Method : PublisherCommonUtils.getOrgId(organizationId)_**

2. In Dev-Portal - Here tenantDomain is an optional request parameter from the user. Thus we need to incorporate that to create an **orgId**. So we need to pass **organizationId** as well as **xWSO2Tenant** to the method logic. If **xWSO2Tenant**  is null we are getting the tenantDomain from the context.

**_Method : RestApiCommonUtil.getOrgIdMatchesTenantDomain(organizationId, xWSO2Tenant)_**

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java (hunk 13 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java:401]: In this method, we are setting the **orgId** as organizationId or tenantDomain. And that tenantDomain is a user given (optional) parameter.

If this method does not signify the logic shall we rename that to one of the following:

1. getOrgId
2. getOrgIdFromTenantDomain
3. retriveOrgIdFromOrganizationIdentifiers

Please suggest a suitable name for this :)

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java (hunk 4 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/20a6fb69a612c320d26f57d1d19519722fe30ab5

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java (hunk 5 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIProvider.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/20a6fb69a612c320d26f57d1d19519722fe30ab5

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/dto/Organization.java (hunk 11 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/dto/Organization.java:42]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/20a6fb69a612c320d26f57d1d19519722fe30ab5

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java (hunk 15 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: It need not be changed. Reverted back to the original version. https://github.com/wso2/carbon-apimgt/pull/9609/commits/20a6fb69a612c320d26f57d1d19519722fe30ab5

**Review by malinthaprasan — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 19]
```diff
@@ -183,8 +183,7 @@ private ApiMgtDAO() {
         PreparedStatement ps = null;
         List<String> versionList = new ArrayList<String>();
         ResultSet resultSet = null;
-
-        String sqlQuery = SQLConstants.GET_VERSIONS_MATCHES_API_NAME_SQL;
+        String sqlQuery = sqlQuery = SQLConstants.GET_VERSIONS_MATCHES_API_NAME_SQL;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:186]: Unnecessary change?

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/006bf23b34ccb736092e59359358e91445012850

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 9 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/006bf23b34ccb736092e59359358e91445012850

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 8 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/006bf23b34ccb736092e59359358e91445012850

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 6 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:1000]: We need the **OrganizationDTO** object when we require both the tenantDomain as well as the **orgId**. 
In the cases where setting an **orgId** into API object, we don't need that extra attribute of tenantDomain.

Is there any special reason to pass an **OrganizationDTO** object instead of **orgId** ?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java (hunk 10 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java:2825]: We can't assign **CARBON.SUPER** to requested tenantDomain. This method is called by Dev-Portal Rest APIs too as well as in all on-prem workflows. In Dev-Portal requests, if a user requests for a tenant domain we need to pass that to be set as the requestedTenantDomain attribute.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 7 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:1806]: We need to set the org Object with either the organizationId from API or with the tenant domain. In a single line, we can't achieve this nah ?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java (hunk 17 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/fa57cf21e6945582abb70eda22a80b00453a4fb7

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java (hunk 12 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.common/src/main/java/org/wso2/carbon/apimgt/rest/api/common/RestApiCommonUtil.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/fa57cf21e6945582abb70eda22a80b00453a4fb7

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java (hunk 18 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/629fd2c956b66218c6f5b9d125e4b0f25fe735e4

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java (hunk 16 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: Fixed https://github.com/wso2/carbon-apimgt/pull/9609/commits/629fd2c956b66218c6f5b9d125e4b0f25fe735e4

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java (hunk 3 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIManager.java:21]: The OrganizationDTO just carries the data from the REST layer to the Persistence layer Nah?  Is it wrong to have that in the DTO section?

In the effort of merging the Organization (in persistence package) and this DTO, we can have it under the model package. Is this making sense?

**Comment by Sarangan0219** [comment]
This PR can be close as these concerns are addressed by https://github.com/wso2/carbon-apimgt/pull/9684 and https://github.com/wso2/carbon-apimgt/pull/9658
